### PR TITLE
[geometry] optimize point distance to point

### PR DESCRIPTION
**Pull request by nirvn**
## Description
A nice little optimization (~6.5%) suggested by @nyalldawson .

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/core/geometry/qgsgeometry.cpp [hunk 1]
```diff
@@ -1595,6 +1595,12 @@ double QgsGeometry::distance( const QgsGeometry &geom ) const
     return -1.0;
   }
 
+  // avoid calling geos for trivial point-to-point distance calculations
+  if ( QgsWkbTypes::flatType( d->geometry->wkbType() ) == QgsWkbTypes::Point && QgsWkbTypes::flatType( geom.constGet()->wkbType() ) == QgsWkbTypes::Point )
```
Comment (m-kuhn) [inline src/core/geometry/qgsgeometry.cpp]: Can `wkbType()` not directly be called on geom? Or is the idea to explicitly skip the nullptr check in `QgsGeometry::wkbType()`?

File: src/core/geometry/qgsgeometry.cpp [hunk 2]
```diff
@@ -1595,6 +1595,12 @@ double QgsGeometry::distance( const QgsGeometry &geom ) const
...
+  {
+    return qgsgeometry_cast< QgsPoint * >( d->geometry.get() )->distance( *qgsgeometry_cast< const QgsPoint * >( geom.constGet() ) );
```
Comment (m-kuhn) [inline src/core/geometry/qgsgeometry.cpp]: `distance()` can be called on a const point as well, so better make it a `constGet()`, no reason to detach.

Comment (nirvn) [reply src/core/geometry/qgsgeometry.cpp]: @m-kuhn , I'm being told by my console that `d->geometry` doesn't have a `constGet()` function.

File: src/core/geometry/qgsgeometry.cpp (hunk 1 above)
Comment (nirvn) [reply src/core/geometry/qgsgeometry.cpp]: Good point.

File: src/core/geometry/qgsgeometry.cpp (hunk 2 above)
Comment (m-kuhn) [reply src/core/geometry/qgsgeometry.cpp]: hah, you are of course right :)

**Pull request merged**